src/benchmarkable.jl

```diff
@@ -27,33 +27,36 @@
 #     (3): n_evals::Integer: The number of times the core expression will be
 #         evaluated per sample.
 
-# recursively add non-constant symbols in module m to a list
-find_nonconsts(m, x::Any, list) = list
-find_nonconsts(m, s::Symbol, list) = isconst(m, s) ? list : push!(list, s)
-function find_nonconsts(m, e::Expr, list)
-    e.head == :line && return # skip line nodes
-    map(a->find_nonconsts(m,a,list), e.args)
-    list
-end
-
 macro benchmarkable(name, setup, core, teardown)
-    inner = gensym(:inner)
-    # Go through the passed core expression and determine non-constant bindings
-    nonconst = unique(find_nonconsts(current_module(), core, Symbol[]))
-    # We assign these non-const bindings to a local constant binding
-    # in order to ensure we aren't timing untyped dispatch lookup
-    nonconst_locals = map(gensym, nonconst)
-    decls = Expr(:block, map((r,s)->:(const $r = $(esc(s))), nonconst_locals, nonconst)...)
+    # expand the passed expression to support `x[1]` and macros
+    expr = expand(core)
+    # only support function calls
+    expr.head == :call || throw(ArgumentError("expression to benchmark must be a function call"))
+    f = expr.args[1]
+    nargs = length(expr.args) - 1
+    args = Symbol[gensym("arg_$i") for i=1:nargs]
+
+    # We create two functions -- an outer function that is a closure around
+    # the expressions used for the arguments the user passed in `core`
+    # and an inner `bench` function that does all the work. This allows the
+    # benchmarked expression's arguments to be considered setup, which are
+    # evaluated in the correct scope, but they're then passed through a function
+    # barrier.  This means that within the `bench` function, these values are
+    # properly inferred and concretely typed.
     quote
-        @noinline function $(inner)($(map(esc, nonconst)...))
-            $(esc(core))
-        end
-        $decls
         function $(esc(name))(
                 s::Samples,
                 n_samples::Integer,
                 evaluations::Integer,
             )
+            bench(s, n_samples, evaluations, $(map(esc, expr.args[2:end])...))
+        end
+        function bench(
```

Comment:
This doesn't need a gensym if you run many different benchmarks on different functions and different argument types in a row?